Reactive Messaging codestart

vgallet opened the pull request:
* fix #20872  
* created `MyReactiveMessagingApplication` to demonstrates `@Emitter`, `@Incoming` and `@Outgoing` annotations.

vgallet committed:
Reactive Messaging codestart

 * fix #20872
 * created `MyReactiveMessagingApplication` to demonstrates `@Emitter`, `@Incoming` and `@Outgoing` annotations.
 * created dynamic application.yml based on selected extension
 * added codestart integration test

vgallet commented:
@ia3andy @ozangunalp 

Please can you take a look at this draft and how I created a dynamic `application.yml`. I also put some `TODO` into the code to show what's missing or if it's necessary to have it.

ia3andy commented on devtools/project-core-extension-codestarts/src/main/resources/codestarts/quarkus/extension-codestarts/reactive-messaging-codestart/codestart.yml:
```diff
@@ -0,0 +1,18 @@
+language:
+  base:
+    dependencies:
+    # TODO: missing HTTP & RabbitMQ connectors ? Not sure to add them as there is no quarkus-smallrye-reactive-messaging extension ?
+      - io.quarkus:quarkus-smallrye-reactive-messaging-kafka
+      - io.quarkus:quarkus-smallrye-reactive-messaging-mqtt
+      - io.quarkus:quarkus-smallrye-reactive-messaging-amqp
```
```suggestion
```

You shouldn't add them manually else they will be added automatically in the generated app. We don't want that :)

ozangunalp replied:
I am guessing that the dependency of the extension is already present in the generated codestart.

ia3andy replied:
@ozangunalp a user is going to select one (or more) of the connectors as extensions, the codestart will be binded to all of them (but deduplicted if more than one is selected). Therefore the selected connectors extensions will be in the pom.xml, no need to add them.

vgallet replied:
fixed

ia3andy reviewed (changes requested):
You also need to add the QuarkusCodestartTest, it will also help you make sure it works fine.

ozangunalp commented on devtools/project-core-extension-codestarts/src/main/resources/codestarts/quarkus/extension-codestarts/reactive-messaging-codestart/base/src/main/resources/application.tpl.qute.yml:
```diff
@@ -0,0 +1,163 @@
+  {#if it.equals("io.quarkus:quarkus-smallrye-reactive-messaging-kafka")}
+    {#if count == 1 }
+
+kafka:
```
It is better to remove this for the dev services to create a Kafka broker

vgallet replied:
fixed

ozangunalp commented on devtools/project-core-extension-codestarts/src/main/resources/codestarts/quarkus/extension-codestarts/reactive-messaging-codestart/base/src/main/resources/application.tpl.qute.yml:
```diff
@@ -0,0 +1,163 @@
+      source-out:
+        connector: smallrye-kafka
+        value:
+          serializer: org.apache.kafka.common.serialization.StringSerializer
```
We can remove Serde altogether as it'll be detected automatically

vgallet replied:
fixed

ozangunalp commented on devtools/project-core-extension-codestarts/src/main/resources/codestarts/quarkus/extension-codestarts/reactive-messaging-codestart/base/src/main/resources/application.tpl.qute.yml:
```diff
@@ -0,0 +1,163 @@
+      source-in:
+        connector: smallrye-kafka
+        value:
+          serializer: org.apache.kafka.common.serialization.StringDeserializer
```
It should be `deserializer`

vgallet replied:
fixed

ozangunalp commented on devtools/project-core-extension-codestarts/src/main/resources/codestarts/quarkus/extension-codestarts/reactive-messaging-codestart/base/src/main/resources/application.tpl.qute.yml:
```diff
@@ -0,0 +1,163 @@
+    {#if it.equals("io.quarkus:quarkus-smallrye-reactive-messaging-amqp")}
+      {#if count == 1 }
+
+amqp-host: localhost
```
Same for amqp dev services.

vgallet replied:
fixed

vgallet commented:
@ia3andy @ozangunalp Hi,

I fixed your comments and added an integration test. Nevertheless, I added three TODO as I'm stuck to make it work and make it right :

  - [Which kafka guide to choose ?](https://github.com/quarkusio/quarkus/pull/21068/files#diff-5a5569128cc474041d28b23055bf226499132af6c814cd45e09a2a68e42933e1R5)
  - [The application.properties file is empty and I don't know why ?](https://github.com/quarkusio/quarkus/pull/21068/files#diff-f0d752ea553ae34c95d9891ebcc1450aa994f72fc19bef19f650cd35bd55d28cR25)
  - [The codestart project cannot be build using an inner class](https://github.com/quarkusio/quarkus/pull/21068/files#diff-f0d752ea553ae34c95d9891ebcc1450aa994f72fc19bef19f650cd35bd55d28cR27)
  
  Thanks for you help

ia3andy commented:
I will check it out to give it a try and see what's wrong.

ia3andy commented:
Ok, I've checked with @mkouba, we can't have a counter in the loop. I think we should not comment (until we find a proper way), the multi-connectors case is an edge case.

I will have another look tomorrow morning with a bit more time, let's wait for until then..

ia3andy commented:
Hey @vgallet, I prepared a commit with the fixes to make it work:
https://github.com/ia3andy/quarkus/commit/154ef45c1161870abdb89455836219a2dd662058

I had to improve a bit the test framework to avoid conflicts between the different tests..

Let me know if you want me to push it to your branch directly..

ia3andy commented:
I had to re-commit due to a problem: https://github.com/ia3andy/quarkus/commit/728964ef613157622085b91cf3244446463b95a7

ia3andy reviewed (approved):
LGTM

ozangunalp commented:
@ia3andy would you know how to test codestarts in local?

ia3andy commented:
@ozangunalp, you can either use the tests, or the CLI:
https://quarkus.io/version/main/guides/extension-codestart#generating

ozangunalp reviewed (approved):
Good for me, Thanks @vgallet !

ia3andy commented:
@vgallet could you run `mvn process-sources` and push?

vgallet commented:
Hello guys @ia3andy @ozangunalp! I discussed with @loicmathieu about the failing tests and he suggested me to split the test `ReactiveMessagingCodestartIT` into three tests, each dedicated per extension. WDYT ?

BTW, it misses the annotation `@EnabledIfSystemProperty(named = "build-projects", matches = "true")` on the test `ReactiveMessagingCodestartIT#buildAll` ?

ia3andy commented:
> Hello guys @ia3andy @ozangunalp! I discussed with @loicmathieu about the failing tests and he suggested me to split the test `ReactiveMessagingCodestartIT` into three tests, each dedicated per extension. WDYT ?
> 

I will try to run the tests locally to see what's going on. I don't think we need to split.

 I
> BTW, it misses the annotation `@EnabledIfSystemProperty(named = "build-projects", matches = "true")` on the test `ReactiveMessagingCodestartIT#buildAll` ?

You don't need it, it your case you want that test to be run.

ia3andy committed:
Improve testId in QuarkusCodestartTest and fix QuarkusCodestartBuildIT

ia3andy commented:
I pushed fixes, let see what the CI has to see to it :)

If it's green, could you squash it all? Thanks @vgallet ✌️

ia3andy commented:
Ok the tests are failing because there is no docker environment on CI for the devservices to be started :-/ I am not sure what's the way forward..

ia3andy commented:
It's weird that the same test is only failing on windows:
```
Error: ]: Build step io.quarkus.kafka.client.deployment.DevServicesKafkaProcessor#startKafkaDevService threw an exception: java.lang.RuntimeException: java.lang.IllegalStateException: Previous attempts to find a Docker environment failed. Will not retry. Please see logs and check configuration
	at io.quarkus.kafka.client.deployment.DevServicesKafkaProcessor.startKafkaDevService(DevServicesKafkaProcessor.java:116)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at io.quarkus.deployment.ExtensionLoader$2.execute(ExtensionLoader.java:887)
	at io.quarkus.builder.BuildContext.run(BuildContext.java:277)
	at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2449)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1478)
	at java.base/java.lang.Thread.run(Thread.java:829)
	at org.jboss.threads.JBossThread.run(JBossThread.java:501)
Caused by: java.lang.IllegalStateException: Previous attempts to find a Docker environment failed. Will not retry. Please see logs and check configuration
	at org.testcontainers.dockerclient.DockerClientProviderStrategy.getFirstValidStrategy(DockerClientProviderStrategy.java:114)

ia3andy commented:
@cescoffier do you know why this test is not working on Windows?

cescoffier commented:
Nope, it should not spend on the OS

loicmathieu commented:
@ia3andy Windows CI didn't have Docker, all IT that needs Docker are not run inside the Windows job.

ia3andy commented:
Should we disable this build test on Windows then?
Or maybe we should disable the `MyReactiveMessagingApplicationTest` on Windows?

loicmathieu commented:
As far as I know, all test that requires Docker are disabled in the CI.

ia3andy commented:
Ok what do we suggest users in that matter? Maybe we should disable this test directly in the template so that users know it doesn't work on Windows? The problem is not related to the codestart test but more the generated project IMO.

ia3andy commented:
@cescoffier what is our best practice for users regarding dev services using docker and Windows?

loicmathieu commented:
@ia3andy Devservices works on Windows, I use them everyday.

But our Windows CI runner didn't have Docker so you must disabled the test via `@DisableOnOs(OS.WINDOWS)`

ia3andy commented:
> @DisableOnOs(OS.WINDOWS)

Thanks, I'll add this then :)

ia3andy committed:
Disable build test on Windows because no docker on CI

ia3andy commented:
@famod any idea why we don't run devtools tests also on JDK 17

ia3andy commented:
I've found the problem and fixed it on my branch, @vgallet could you rebase from main so I can apply it?

famod commented:
@ia3andy
> @famod any idea why we don't run devtools tests also on JDK 17

Well, there simply wasn't a Java 16 job when I updated to 17. See also: https://github.com/quarkusio/quarkus/issues/21481#issuecomment-970472454

CI so many jobs already, but as devtools ITs "only" take ~30m I think this could be added.

cc @gsmet

ia3andy commented:
I am closing this one to be able to rebase and test with the CI (I am keeping you @vgallet as author): https://github.com/quarkusio/quarkus/pull/21906
